cmake/JuliaUtils.cmake

```diff
@@ -0,0 +1,88 @@
+#
+
+include(CMakeVarMacros)
+
+get_filename_component(jl_cmake_utils_dir
+  "${CMAKE_CURRENT_LIST_FILE}" PATH)
+
+set(jl_cmake_utils_dir "${jl_cmake_utils_dir}" CACHE "" INTERNAL FORCE)
+
+function(jl_std_fpath var path)
+  if(NOT IS_ABSOLUTE "${path}")
+    set(${var} "${path}" PARENT_SCOPE)
+    return()
+  endif()
+  file(RELATIVE_PATH src_path "${CMAKE_CURRENT_SOURCE_DIR}" "${path}")
+  file(RELATIVE_PATH bin_path "${CMAKE_CURRENT_BINARY_DIR}" "${path}")
+  string(LENGTH "${src_path}" src_len)
+  string(LENGTH "${bin_path}" bin_len)
+  if(src_len GREATER bin_len)
+    set(${var} "${bin_path}" PARENT_SCOPE)
+  else()
+    set(${var} "${src_path}" PARENT_SCOPE)
+  endif()
+endfunction()
+
+function(jl_fpath_to_target var file)
+  jl_std_fpath(std_fname "${file}")
+  string(REGEX REPLACE "[^-_.a-zA-Z0-9]" "_" std_fname "${std_fname}")
+  string(REGEX REPLACE "_+" "_" std_fname "${std_fname}")
+  cmake_utils_get_unique_name("${std_fname}" target)
+  set("${var}" "${target}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_dep_list var deps)
+  set(new_deps)
+  foreach(dep ${deps})
+    if(TARGET "${dep}")
+      set(new_deps ${new_deps} "$<TARGET_FILE:${dep}>")
+    elseif(IS_ABSOLUTE "${dep}")
+      set(new_deps ${new_deps} ${dep})
+    else()
+      get_filename_component(dep "${dep}" ABSOLUTE)
+      set(new_deps ${new_deps} ${dep})
+    endif()
+  endforeach()
+  set(${var} "${new_deps}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_output_list var outputs)
+  set(new_outputs)
+  foreach(output ${outputs})
+    if(IS_ABSOLUTE "${output}")
+      set(new_outputs ${new_outputs} ${output})
+    else()
+      set(new_outputs ${new_outputs} "${CMAKE_CURRENT_BINARY_DIR}/${output}")
+    endif()
+  endforeach()
+  set(${var} "${new_outputs}" PARENT_SCOPE)
+endfunction()
+
+function(jl_custom_target target outputs output_deps target_deps autodep_file)
+  jl_rewrite_dep_list(output_deps "${output_deps}")
+  jl_rewrite_output_list(outputs "${outputs}")
+  if((NOT "${autodep_file}" STREQUAL "") AND
+      (NOT IS_ABSOLUTE "${autodep_file}"))
+    set(autodep_file "${CMAKE_CURRENT_BINARY_DIR}/${autodep_file}")
+  endif()
+  add_custom_target("${target}" ALL
+    COMMAND "${CMAKE_COMMAND}"
+    "-DOUTPUTS=${outputs}" "-DOUTPUT_DEPS=${output_deps}"
+    "-DAUTODEP_FILE=${autodep_file}" "-DCOMMAND_SPEC=${ARGN}"
+    "-DCUR_BIN_DIR=${CMAKE_CURRENT_BINARY_DIR}"
+    "-DCUR_SRC_DIR=${CMAKE_CURRENT_SOURCE_DIR}"
+    -P "${jl_cmake_utils_dir}/jl_custom_target.cmake"
+    DEPENDS ${target_deps} VERBATIM)
+endfunction()
```

Comment:
> We do because the sysimg need to be recompiled?

We don't need to re-configure, but we would need to re-execute the custom command. As long as the globbed dependencies get reflected in the generated makefiles I suspect it should be fine, but I'll need to do some tests to be sure. There's no way currently of adding or removing a julia file from the system image without modifying at least one existing julia file that previously went into the system image. So the configure-time glob results should still be usable as the make-time dependencies that would trigger the custom command, I believe.

The main thing that I don't like about this is it's nonstandard and difficult to understand. Maybe the latter could be addressed by copious comments, but if you need copious comments to make something understandable that's already a bad sign. You have a much better understanding of what this is doing than anyone else can, you wrote it. I'm still very lost after trying to read this code several times.